sdk/lib/drivers/sptilib/sptilib.c

```diff
@@ -0,0 +1,1042 @@
+/*
+ * PROJECT:     ReactOS Storage Stack
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     ATA and SCSI Pass Through Interface for storage drivers
+ * COPYRIGHT:   Copyright 2025 Dmitry Borisov <di.sean@protonmail.com>
+ */
+
+/*
+ * This library provides helper code for handling ATA and SCSI Pass Through IOCTLs.
+ * Typically, these IRPs come from user-mode applications.
+ * The handler will translate the IOCTL into an SRB and send it to the PDO.
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <ntddk.h>
+#include <ntintsafe.h>
+#include <pseh/pseh2.h>
+#include <scsi.h>
+#include <ntddscsi.h>
+
+#define NDEBUG
+#include <debug.h>
+
+#include "sptilib.h"
+#include "sptilibp.h"
+
+/* PRIVATE FUNCTIONS **********************************************************/
+
+_At_(IrpContext->Srb.SenseInfoBuffer, __drv_freesMem(Mem))
+static
+CODE_SEG("PAGE")
+VOID
+PassthruFreeIrpContext(
+    _In_opt_ __drv_freesMem(Mem) PPASSTHROUGH_IRP_CONTEXT IrpContext)
+{
+    PIRP Irp;
+
+    PAGED_CODE();
+
+    if (!IrpContext)
+        return;
+
+    Irp = IrpContext->Irp;
+    if (Irp)
+    {
+        if (Irp->MdlAddress)
+        {
+            MmUnlockPages(Irp->MdlAddress);
+            IoFreeMdl(Irp->MdlAddress);
+        }
+        Irp->MdlAddress = NULL;
+        IoFreeIrp(Irp);
+    }
+
+    if (IrpContext->Srb.SenseInfoBuffer)
+        ExFreePoolWithTag(IrpContext->Srb.SenseInfoBuffer, TAG_SPTI);
+
+    ExFreePoolWithTag(IrpContext, TAG_SPTI);
+}
+
+static IO_COMPLETION_ROUTINE PassthruCompletion;
+static
+NTSTATUS
+NTAPI
+PassthruCompletion(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp,
+    _In_reads_opt_(_Inexpressible_("varies")) PVOID Context)
+{
+    UNREFERENCED_PARAMETER(DeviceObject);
+
+    if (Irp->PendingReturned)
+        KeSetEvent(Context, IO_NO_INCREMENT, FALSE);
+
+    return STATUS_MORE_PROCESSING_REQUIRED;
+}
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+PassthruCreateIrpContext(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP OriginalIrp,
+    _In_ PVOID DataBuffer,
+    _In_ ULONG DataBufferLength,
+    _In_ BOOLEAN IsDirectMemoryAccess,
+    _In_ BOOLEAN IsBufferReadAccess,
+    _Outptr_ __drv_allocatesMem(Mem) PPASSTHROUGH_IRP_CONTEXT* Result)
+{
+    PPASSTHROUGH_IRP_CONTEXT IrpContext;
+    PIRP Irp;
+    PIO_STACK_LOCATION IoStack;
+
+    PAGED_CODE();
+
+    *Result = IrpContext = ExAllocatePoolZero(NonPagedPool, sizeof(*IrpContext), TAG_SPTI);
```

Comment:
Would it be wise to defer setting anything inside `*Result` up to until you are sure `PassthruCreateIrpContext` is going to return success?
Getting a meaningful output, e.g. `Result` (especially an allocation) while the function (e.g. `PassthruCreateIrpContext`) fails with an error code, is unnatural.

Reply:
> [...] since SptiFreeIrpContext should free the memory if PassthruCreateIrpContext fails.

If PassthruCreateIrpContext fails, why would you expect SptiFreeIrpContext to do something? (This sounds like, "ExFreePoolWithTag should free the buffer if ExAllocatePoolWithTag fails" , or, "NtClose should close the handle if NtOpenFile fails :thonk: :thonk:)